compiler/src/dotty/tools/debug/ResolveReflectEval.scala

```diff
@@ -0,0 +1,346 @@
+package dotty.tools.debug
+
+import dotty.tools.dotc.core.SymUtils
+import dotty.tools.dotc.ast.tpd.*
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts.*
+import dotty.tools.dotc.core.Flags.*
+import dotty.tools.dotc.core.NameKinds.QualifiedInfo
+import dotty.tools.dotc.core.Names.*
+import dotty.tools.dotc.core.Phases
+import dotty.tools.dotc.core.StdNames.*
+import dotty.tools.dotc.core.Symbols.*
+import dotty.tools.dotc.core.TypeErasure.ErasedValueType
+import dotty.tools.dotc.core.Types.*
+import dotty.tools.dotc.report
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+import dotty.tools.dotc.transform.ValueClasses
+
+private class ResolveReflectEval(config: ExpressionCompilerConfig, expressionStore: ExpressionStore) extends MiniPhase:
+  override def phaseName: String = ResolveReflectEval.name
+
+  override def transformTypeDef(tree: TypeDef)(using Context): Tree =
+    ExpressionTransformer.transform(tree)
+
+  object ExpressionTransformer extends TreeMap:
+    override def transform(tree: Tree)(using Context): Tree =
+      tree match
+        case tree: DefDef if tree.symbol == config.evaluateMethod =>
+          // unbox the result of the `evaluate` method if it is a value class
+          val gen = new Gen(
+            Apply(
+              Select(This(config.expressionClass), termName("reflectEval")),
+              List(nullLiteral, nullLiteral, nullLiteral)
+            )
+          )
+          val rhs = gen.unboxIfValueClass(expressionStore.symbol.nn, transform(tree.rhs))
+          cpy.DefDef(tree)(rhs = rhs)
+
+        case reflectEval: Apply if isReflectEval(reflectEval.fun.symbol) =>
+          val qualifier :: _ :: argsTree :: Nil = reflectEval.args.map(transform): @unchecked
+          val args = argsTree.asInstanceOf[JavaSeqLiteral].elems
+          val gen = new Gen(reflectEval)
+          tree.attachment(ReflectEvalStrategy) match
+            case ReflectEvalStrategy.This(cls) => gen.getThisObject
+            case ReflectEvalStrategy.LocalOuter(cls) => gen.getLocalValue("$outer")
+            case ReflectEvalStrategy.Outer(outerCls) => gen.getOuter(qualifier, outerCls)
+            case ReflectEvalStrategy.LocalValue(variable, isByName) =>
```

Comment:
This match is pretty dense. For readability, consider adding a blank line between multi-line `case`s. That helps the brain's chunking mechanism.

Reply:
done